kratos/linear_solvers/amgcl_solver_impl.hpp

```diff
@@ -0,0 +1,763 @@
+            default_parameters.AddInt("block_size", 1);
+        } else if (block_size_settings.Is<std::string>()) {
+            KRATOS_ERROR_IF_NOT(block_size_settings.Get<std::string>() == "auto")
+                << "Invalid value for \"block_size\": \"" << block_size_settings.Get<std::string>() << "\". "
+                << "Expecting a positive integer, or \"auto\".";
+            default_parameters.AddString("block_size", "auto");
+        } else {
+            KRATOS_ERROR << "Invalid type for \"block_size\". Expecting a positive integer, or \"auto\".";
+        }
+    }
+    Settings.ValidateAndAssignDefaults(default_parameters);
+    KRATOS_CATCH("")
+
+    KRATOS_TRY
+    CheckIfSelectedOptionIsAvailable(Settings,
+                                     "preconditioner_type",
+                                     {
+                                        "amg",
+                                        "relaxation",
+                                        "dummy"
+                                     });
+    KRATOS_CATCH("")
+
+    KRATOS_TRY
+    const std::string preconditioner_type = Settings["preconditioner_type"].GetString();
+    mAMGCLParameters.put("precond.class", preconditioner_type);
+    if (preconditioner_type == "relaxation")
+        this->SetSmootherType(Settings["smoother_type"].GetString());
+
+    mVerbosity=Settings["verbosity"].GetInt();
+    mGMRESSize = Settings["gmres_krylov_space_dimension"].GetInt();
+    mUseBlockMatricesIfPossible = Settings["use_block_matrices_if_possible"].GetBool();
+
+    // Termination criteria.
+    mTolerance = Settings["tolerance"].GetDouble();
+    mMaxIterationsNumber = Settings["max_iteration"].GetInt();
+    mAMGCLParameters.put("solver.tol", mTolerance);
+    mAMGCLParameters.put("solver.maxiter", mMaxIterationsNumber);
+
+    mFallbackToGMRES = false;
+    this->SetIterativeSolverType(Settings["krylov_type"].GetString());
+
+    // Multigrid preconditioner settings.
+    mProvideCoordinates = Settings["provide_coordinates"].GetBool();
+    mCoarseEnough = Settings["coarse_enough"].GetInt();
+    mUseAMGPreconditioning = preconditioner_type == "amg";
+
+    if(mUseAMGPreconditioning) {
+        this->SetSmootherType(Settings["smoother_type"].GetString());
+        this->SetCoarseningType(Settings["coarsening_type"].GetString());
+
+        const int max_levels = Settings["max_levels"].GetInt();
+        if(max_levels >= 0) mAMGCLParameters.put("precond.max_levels",  max_levels);
+
+        mAMGCLParameters.put("precond.npre",  Settings["pre_sweeps"].GetInt());
+        mAMGCLParameters.put("precond.npost",  Settings["post_sweeps"].GetInt());
+    } // is mUseAMGPreconditioning
+
+    // GPU settings.
+    mUseGPGPU = Settings["use_gpgpu"].GetBool();
+    if (mUseGPGPU) {
+        // ILU0 in a GPU backend has approximate iterative implementation.
+        // Increase the default number of iterations to make ILU0 more robust.
+        int ilu0_iters = 9;
```

Comment:
Woudn't it be better to take this from the user, and `9` to be the default for the setting?